### PR TITLE
A few helpers

### PR DESCRIPTION
**Pull request by pelikhan**
- [x] cleanup debug view
- [x] a few top-level helpers

Here is a test makecode program that can be used in the pxt-maker editor. Use the ``jacdac-ts`` branches for pxt, pxt-common-packages, pxt-maker.

* ``pxt serve --rebundle``
* open CPX
* add this code

```
jacdac.consolePriority = ConsolePriority.Log;
jacdac.consoleService.start();
jacdac.consoleService.setConsoleMode(JDConsoleMode.Listen);
jacdac.buttonAService.start();
jacdac.stop()
jacdac.start();

forever(function () {
    console.log(`connected: ${jacdac.isConnected()}`)
    console.log(`running: ${jacdac.isRunning()}`)
    console.log(`enumerated: ${jacdac.JACDAC.instance.controlService.isEnumerated()}`)
    console.log(`enumerating: ${jacdac.JACDAC.instance.controlService.isEnumerating()}`)
    const devices = jacdac.devices();
    console.log(`${devices.length} devices`)
    devices.forEach(function (value: jacdac.JDDevice, index: number) {
        console.log(value.toString());
    })
    pause(5000)
})
```

Current output:
```
l>--:ff>log>start

common-sim.js:591 l>--:ff>log>mode 2

common-sim.js:591 l>--:ff>buttonA>start

common-sim.js:591 l>connected: false

common-sim.js:591 l>running: true

common-sim.js:591 l>enumerated: false

common-sim.js:591 l>enumerating: false

common-sim.js:591 l>0 devices
```

### CODE REVIEW

File: source/jacdac.ts [hunk 1]
```diff
@@ -164,4 +172,26 @@ namespace jacdac {
...
+    export function devices(): JDDevice[] {
+        return jacdac.JACDAC.instance.devices;
+    }
+
+    /**
+     * Gets a value that indicates if the JACDAC bus is connected.
+     */
+    export function isConnected(): boolean {
+        return jacdac.JACDAC.instance.bus.isConnected()
+            && jacdac.JACDAC.instance.controlService.isConnected();
```
Comment (pelikhan) [inline source/jacdac.ts]: double check this

File: source/structs.ts [hunk 2]
```diff
@@ -192,14 +196,18 @@ namespace jacdac {
...
                 this.device_address = 0;
                 this.device_flags = 0;
-                this.services = []
+                this._services = []
                 this.rolling_counter = 0;
             }
         }
 
+        get services(): JDServiceInformation[] {
+            return this._services ? this._services.slice(0) : [];
```
Comment (pelikhan) [inline source/structs.ts]: can _services be null/undefined?

File: source/jacdac.ts (hunk 1 above)
Comment (jamesadevine) [reply source/jacdac.ts]: 👍

File: source/structs.ts [hunk 3]
```diff
@@ -64,6 +64,10 @@ namespace jacdac {
         getBuffer() : Buffer {
             return this._buffer;
         }
+
+        toString(): string {
+            return this._buffer.toHex();
```
Comment (jamesadevine) [inline source/structs.ts]: Do we want this overhead?

File: source/structs.ts (hunk 2 above)
Comment (jamesadevine) [reply source/structs.ts]: I don't think so, you can see above it is set to _services

Comment (jamesadevine) [reply source/structs.ts]: this._services = []

File: source/structs.ts (hunk 3 above)
Comment (pelikhan) [reply source/structs.ts]: It matches the makecode impl. No overhead until you call it.

File: source/structs.ts [hunk 4]
```diff
@@ -230,6 +238,10 @@ namespace jacdac {
         get udid() {
             return this.unique_device_identifier;
         }
+
+        toString(): string {
+            return `${this.udid.toHex()}:${this.device_address}:${this.device_name || "--"}`;
```
Comment (jamesadevine) [inline source/structs.ts]: Shouldn't we just have a debug view for a Packet and a device?

File: source/structs.ts (hunk 3 above)
Comment (jamesadevine) [reply source/structs.ts]: If it's unused does it still get compiled in?

File: source/control.ts [hunk 5]
```diff
@@ -169,11 +175,13 @@ namespace jacdac{
                 hostServiceCount++;
             }
 
-            if (hostServiceCount)
-            {
+            // this optimization is bogus because you will call enumerate again
+            // and again until a host service is loaded
+            //if (hostServiceCount)
+            //{
                 this.status |= (JD_CONTROL_SERVICE_STATUS_ENUMERATING | JD_CONTROL_SERVICE_STATUS_ENUMERATE);
```
Comment (pelikhan) [inline source/control.ts]: i had to comment this out to make it work.

File: source/devicemanager.ts [hunk 6]
```diff
@@ -7,7 +7,7 @@ namespace jacdac {
         }
 
         getDeviceList(): JDDevice[] {
-            return this.devices.slice(0);
+            return this.devices.slice(0, this.devices.length);
```
Comment (pelikhan) [inline source/devicemanager.ts]: workaround pxt bug, now fix so can be reverted

File: source/structs.ts (hunk 3 above)
Comment (pelikhan) [reply source/structs.ts]: no compiled away

File: source/structs.ts (hunk 4 above)
Comment (pelikhan) [reply source/structs.ts]: it's just a convienice to have a toString. Once we have our debugger in place, you will see this string in the editor.

**Comment by jamesadevine** [comment]
TODO tomorrow. Add options hooks for: name change and identification request

**Pull request merged**